File: src/main/java/com/epri/dss/delivery/impl/GICTransformerObjImpl.java
package com.epri.dss.delivery.impl;

import java.io.PrintStream;

import com.epri.dss.common.DSSClass;
import com.epri.dss.delivery.GICTransformer;
import com.epri.dss.delivery.GICTransformerObj;
import com.epri.dss.parser.impl.Parser;
import com.epri.dss.shared.CMatrix;
import com.epri.dss.shared.impl.CMatrixImpl;
import com.epri.dss.shared.impl.Complex;

public class GICTransformerObjImpl extends PDElementImpl implements GICTransformerObj {

	private double G1, G2;  // single G per phase (line rating)

    private int specType;

	public GICTransformerObjImpl(DSSClass parClass, String transformerName) {
		super(parClass);
		objType = parClass.getDSSClassType();
		setName(transformerName.toLowerCase());

		setNPhases(3);  // directly set conds and phases
		nConds = 3;
		setNTerms(2);  // force allocation of terminals and conductors

		setBus(2, getBus(1) + ".0");  // default to grounded
		isShunt = true;

		G1            = 10000.0;
		G2            = 10000.0;
		specType      = GICTransformer.SPEC_GSU;

		normAmps  = 0.0;
		emergAmps = 0.0;
		faultRate = 0.0;
		pctPerm   = 100.0;
		hrsToRepair = 0.0;

		initPropertyValues(0);

		YOrder = nTerms * nConds;
		recalcElementData();
	}

	@Override
	public void recalcElementData() {
		// nothing to do
	}

	@Override
	public void calcYPrim() {
		Complex value, value2;
		int i;

		CMatrix YPrimTemp;

		if (isYprimInvalid()) {  // reallocate YPrim if something has invalidated old allocation
			if (YPrimSeries != null) YPrimSeries = null;
			YPrimSeries = new CMatrixImpl(YOrder);
			if (YPrimShunt != null) YPrimShunt = null;
			YPrimShunt = new CMatrixImpl(YOrder);
			if (YPrim != null) YPrim = null;
			YPrim = new CMatrixImpl(YOrder);
		} else {
			YPrimSeries.clear();  // zero out YPrim
			YPrimShunt.clear();   // zero out YPrim
			YPrim.clear();
		}

		if (isShunt()) {
			YPrimTemp = YPrimShunt;
		} else {
			YPrimTemp = YPrimSeries;
		}

		// make sure randomMult is 1.0 if not solution mode MonteFault

		/* Now, put in Yprim matrix */

		/* If the fault is not on, the set zero conductance */

		switch (specType) {
		case GICTransformer.SPEC_GSU:

			value = new Complex(G1, 0.0);
			value2 = value.negate();
			for (i = 0; i < nPhases; i++) {
				YPrimTemp.setElement(i, i, value);  // elements are only on the diagonals
				YPrimTemp.setElement(i+nPhases, i+nPhases,value);
				YPrimTemp.setElemSym(i, i+nPhases, value2);
			}
			break;

		case GICTransformer.SPEC_AUTO:

			// terminals 1 and 2
			value = new Complex(G1, 0.0);
			value2 = value.negate();
			for (i = 0; i < nPhases; i++) {
				YPrimTemp.setElement(i, i, value);  // elements are only on the diagonals
				YPrimTemp.setElement(i+nPhases, i+nPhases, value);
				YPrimTemp.setElemSym(i, i+nPhases, value2);
			}
			// terminals 3 and 4
			value = new Complex(G2, 0.0);
			value2 = value.negate();
			for (i = 2*nPhases+1; i < 3*nPhases; i++) {
				YPrimTemp.setElement(i, i, value);  // elements are only on the diagonals
				YPrimTemp.setElement(i+nPhases, i+nPhases, value);
				YPrimTemp.setElemSym(i, i+nPhases, value2);
			}
			break;

		case GICTransformer.SPEC_YY:

			// terminals 1 and 2
			value = new Complex(G1, 0.0);
			value2 = value.negate();
			for (i = 0; i < nPhases; i++) {
				YPrimTemp.setElement(i, i, value);  // elements are only on the diagonals
				YPrimTemp.setElement(i+nPhases, i+nPhases, value);
				YPrimTemp.setElemSym(i, i+nPhases, value2);
			}
			// terminals 3 and 4
			value = new Complex(G2, 0.0);
			value2 = value.negate();
			for (i = 2*nPhases+1; i < 3*nPhases; i++) {
				YPrimTemp.setElement(i, i, value);  // elements are only on the diagonals
				YPrimTemp.setElement(i+nPhases, i+nPhases, value);
				YPrimTemp.setElemSym(i, i+nPhases, value2);
			}
			break;
		}

		YPrim.copyFrom(YPrimTemp);

		super.calcYPrim();
		setYPrimInvalid(false);
	}

	@Override
	public void dumpProperties(PrintStream f, boolean complete) {
		int i;

	    super.dumpProperties(f, complete);

	    DSSClass pc = getParentClass();

	    f.println("~ " + pc.getPropertyName()[0]+"="+getFirstBus());
	    f.println("~ " + pc.getPropertyName()[1]+"="+getNextBus());
	    f.println("~ " + pc.getPropertyName()[2]+"="+getNextBus());
	    f.println("~ " + pc.getPropertyName()[3]+"="+getNextBus());

	    f.println("~ " + pc.getPropertyName()[4]+"="+nPhases);
	    switch (specType) {
		case GICTransformer.SPEC_GSU:
			f.println("~ " + pc.getPropertyName()[6] + "= GSU");
			break;
		case GICTransformer.SPEC_AUTO:
			f.println("~ " + pc.getPropertyName()[6] + "= AUTO");
			break;
		case GICTransformer.SPEC_YY:
			f.println("~ " + pc.getPropertyName()[6] + "= YY");
			break;
	    }
	    f.println("~ " + pc.getPropertyName()[6]+"="+(1.0/G1));
	    f.println("~ " + pc.getPropertyName()[7]+"="+(1.0/G2));

	    for (i = GICTransformer.NumPropsThisClass; i < pc.getNumProperties(); i++)
	    	f.println("~ " + pc.getPropertyName()[i]+"="+getPropertyValue(i));
	}

	@Override
	public void initPropertyValues(int arrayOffset) {

		setPropertyValue(0, getBus(1));
		setPropertyValue(1, getBus(2));
		setPropertyValue(2, getBus(3));
		setPropertyValue(3, getBus(4));
		setPropertyValue(4, "3");
		setPropertyValue(5, "GSU");
		setPropertyValue(6, "0.0001");
		setPropertyValue(7, "0.0001");

		super.initPropertyValues(GICTransformer.NumPropsThisClass - 1);

		// Override Inherited Properties
		setPropertyValue(GICTransformer.NumPropsThisClass + 0, "0");  // normAmps
		setPropertyValue(GICTransformer.NumPropsThisClass + 1, "0");  // emergAmps
		setPropertyValue(GICTransformer.NumPropsThisClass + 2, "0");  // faultRate
		setPropertyValue(GICTransformer.NumPropsThisClass + 3, "0");  // pctPerm
		setPropertyValue(GICTransformer.NumPropsThisClass + 4, "0");  // hrsToRepair
	}

	@Override
	public String getPropertyValue(int index) {

		switch (index) {
		case 0:
			return getBus(1);
		case 1:
			return getBus(2);
		case 2:
			return getBus(3);
		case 3:
			return getBus(4);
		case 4:
			return String.format("%d", nPhases);
		case 6:
			return String.format("%.8g", 1.0 / G1);
		case 7:
			return String.format("%.8g", 1.0 / G2);
		default:
			return super.getPropertyValue(index);
		}
	}

	@Override
	public void makePosSequence() {
		if (nPhases != 1) {
			Parser.getInstance().setCmdString("Phases=1");
		    edit();
		}
		super.makePosSequence();
	}

	// FIXME: Private members in OpenDSS

	public double getG1() {
		return G1;
	}

	public void setG1(double g1) {
		G1 = g1;
	}

	public double getG2() {
		return G2;
	}

	public void setG2(double g2) {
		G2 = g2;
	}

	public int getSpecType() {
		return specType;
	}

	public void setSpecType(int type) {
		specType = type;
	}

}

File: pydss/delivery/GICTransformerObjImpl.py
from dss.delivery.impl.PDElementImpl import (PDElementImpl,)
from dss.delivery.GICTransformerObj import (GICTransformerObj,)
from dss.delivery.GICTransformer import (GICTransformer,)
from dss.parser.impl.Parser import (Parser,)
from dss.shared.impl.CMatrixImpl import (CMatrixImpl,)
# from org.apache.commons.math.complex.Complex import (Complex,)


class GICTransformerObjImpl(PDElementImpl, GICTransformerObj):
    G1 = None
    G2 = None
    # single G per phase (line rating)
    specType = None

    def __init__(self, parClass, transformerName):
        super(GICTransformerObjImpl, self)(parClass)
        self.objType = parClass.getDSSClassType()
        self.setName(transformerName.toLowerCase())
        self.setNPhases(3)
        # directly set conds and phases
        self.nConds = 3
        self.setNTerms(2)
        # force allocation of terminals and conductors
        self.setBus(2, self.getBus(0) + '.0')
        # default to grounded
        self.isShunt = True
        self.G1 = 10000.0
        self.G2 = 10000.0
        self.specType = GICTransformer.SPEC_GSU
        self.normAmps = 0.0
        self.emergAmps = 0.0
        self.faultRate = 0.0
        self.pctPerm = 100.0
        self.hrsToRepair = 0.0
        self.initPropertyValues(0)
        self.YOrder = self.nTerms * self.nConds
        self.recalcElementData()

    def recalcElementData(self):
        # nothing to do
        pass

    def calcYPrim(self):
        if self.isYprimInvalid():
            # reallocate YPrim if something has invalidated old allocation
            if self.YPrimSeries is not None:
                self.YPrimSeries = None
            self.YPrimSeries = CMatrixImpl(self.YOrder)
            if self.YPrimShunt is not None:
                self.YPrimShunt = None
            self.YPrimShunt = CMatrixImpl(self.YOrder)
            if self.YPrim is not None:
                self.YPrim = None
            self.YPrim = CMatrixImpl(self.YOrder)
        else:
            self.YPrimSeries.clear()
            # zero out YPrim
            self.YPrimShunt.clear()
            # zero out YPrim
            self.YPrim.clear()
        if self.isShunt():
            YPrimTemp = self.YPrimShunt
        else:
            YPrimTemp = self.YPrimSeries
        # make sure randomMult is 1.0 if not solution mode MonteFault
        # Now, put in Yprim matrix
        # If the fault is not on, the set zero conductance
        _0 = self.specType
        _1 = False
        while True:
            if _0 == GICTransformer.SPEC_GSU:
                _1 = True
                value = Complex(self.G1, 0.0)
                value2 = value.negate()
                _2 = True
                i = 0
                while True:
                    if _2 is True:
                        _2 = False
                    else:
                        i += 1
                    if not (i < self.nPhases):
                        break
                    YPrimTemp.setElement(i, i, value)
                    # elements are only on the diagonals
                    YPrimTemp.setElement(i + self.nPhases, i + self.nPhases, value)
                    YPrimTemp.setElemSym(i, i + self.nPhases, value2)
                break
            if (_1 is True) or (_0 == GICTransformer.SPEC_AUTO):
                _1 = True
                value = Complex(self.G1, 0.0)
                value2 = value.negate()
                _3 = True
                i = 0
                while True:
                    if _3 is True:
                        _3 = False
                    else:
                        i += 1
                    if not (i < self.nPhases):
                        break
                    YPrimTemp.setElement(i, i, value)
                    # elements are only on the diagonals
                    YPrimTemp.setElement(i + self.nPhases, i + self.nPhases, value)
                    YPrimTemp.setElemSym(i, i + self.nPhases, value2)
                # terminals 3 and 4
                value = Complex(self.G2, 0.0)
                value2 = value.negate()
                _4 = True
                i = (2 * self.nPhases) + 1
                while True:
                    if _4 is True:
                        _4 = False
                    else:
                        i += 1
                    if not (i < 3 * self.nPhases):
                        break
                    YPrimTemp.setElement(i, i, value)
                    # elements are only on the diagonals
                    YPrimTemp.setElement(i + self.nPhases, i + self.nPhases, value)
                    YPrimTemp.setElemSym(i, i + self.nPhases, value2)
                break
            if (_1 is True) or (_0 == GICTransformer.SPEC_YY):
                _1 = True
                value = Complex(self.G1, 0.0)
                value2 = value.negate()
                _5 = True
                i = 0
                while True:
                    if _5 is True:
                        _5 = False
                    else:
                        i += 1
                    if not (i < self.nPhases):
                        break
                    YPrimTemp.setElement(i, i, value)
                    # elements are only on the diagonals
                    YPrimTemp.setElement(i + self.nPhases, i + self.nPhases, value)
                    YPrimTemp.setElemSym(i, i + self.nPhases, value2)
                # terminals 3 and 4
                value = Complex(self.G2, 0.0)
                value2 = value.negate()
                _6 = True
                i = (2 * self.nPhases) + 1
                while True:
                    if _6 is True:
                        _6 = False
                    else:
                        i += 1
                    if not (i < 3 * self.nPhases):
                        break
                    YPrimTemp.setElement(i, i, value)
                    # elements are only on the diagonals
                    YPrimTemp.setElement(i + self.nPhases, i + self.nPhases, value)
                    YPrimTemp.setElemSym(i, i + self.nPhases, value2)
                break
            break
        self.YPrim.copyFrom(YPrimTemp)
        super(GICTransformerObjImpl, self).calcYPrim()
        self.setYPrimInvalid(False)

    def dumpProperties(self, f, complete):
        super(GICTransformerObjImpl, self).dumpProperties(f, complete)
        pc = self.getParentClass()
        print '~ ' + pc.getPropertyName()[0] + '=' + self.getFirstBus()
        print '~ ' + pc.getPropertyName()[1] + '=' + self.getNextBus()
        print '~ ' + pc.getPropertyName()[2] + '=' + self.getNextBus()
        print '~ ' + pc.getPropertyName()[3] + '=' + self.getNextBus()
        print '~ ' + pc.getPropertyName()[4] + '=' + self.nPhases
        _0 = self.specType
        _1 = False
        while True:
            if _0 == GICTransformer.SPEC_GSU:
                _1 = True
                print '~ ' + pc.getPropertyName()[6] + '= GSU'
                break
            if (_1 is True) or (_0 == GICTransformer.SPEC_AUTO):
                _1 = True
                print '~ ' + pc.getPropertyName()[6] + '= AUTO'
                break
            if (_1 is True) or (_0 == GICTransformer.SPEC_YY):
                _1 = True
                print '~ ' + pc.getPropertyName()[6] + '= YY'
                break
            break
        print '~ ' + pc.getPropertyName()[6] + '=' + (1.0 / self.G1)
        print '~ ' + pc.getPropertyName()[7] + '=' + (1.0 / self.G2)
        _2 = True
        i = GICTransformer.NumPropsThisClass
        while True:
            if _2 is True:
                _2 = False
            else:
                i += 1
            if not (i < pc.getNumProperties()):
                break
            print '~ ' + pc.getPropertyName()[i] + '=' + self.getPropertyValue(i)

    def initPropertyValues(self, arrayOffset):
        self.setPropertyValue(0, self.getBus(0))
        self.setPropertyValue(1, self.getBus(1))
        self.setPropertyValue(2, self.getBus(2))
        self.setPropertyValue(3, self.getBus(3))
        self.setPropertyValue(4, '3')
        self.setPropertyValue(5, 'GSU')
        self.setPropertyValue(6, '0.0001')
        self.setPropertyValue(7, '0.0001')
        super(GICTransformerObjImpl, self).initPropertyValues(GICTransformer.NumPropsThisClass - 1)
        # Override Inherited Properties
        self.setPropertyValue(GICTransformer.NumPropsThisClass + 0, '0')
        # normAmps
        self.setPropertyValue(GICTransformer.NumPropsThisClass + 1, '0')
        # emergAmps
        self.setPropertyValue(GICTransformer.NumPropsThisClass + 2, '0')
        # faultRate
        self.setPropertyValue(GICTransformer.NumPropsThisClass + 3, '0')
        # pctPerm
        self.setPropertyValue(GICTransformer.NumPropsThisClass + 4, '0')
        # hrsToRepair

    def getPropertyValue(self, index):
        _0 = index
        _1 = False
        while True:
            if _0 == 0:
                _1 = True
                return self.getBus(0)
            if (_1 is True) or (_0 == 1):
                _1 = True
                return self.getBus(1)
            if (_1 is True) or (_0 == 2):
                _1 = True
                return self.getBus(2)
            if (_1 is True) or (_0 == 3):
                _1 = True
                return self.getBus(3)
            if (_1 is True) or (_0 == 4):
                _1 = True
                return String.format.format('%d', self.nPhases)
            if (_1 is True) or (_0 == 6):
                _1 = True
                return String.format.format('%.8g', 1.0 / self.G1)
            if (_1 is True) or (_0 == 7):
                _1 = True
                return String.format.format('%.8g', 1.0 / self.G2)
            if True:
                _1 = True
                return super(GICTransformerObjImpl, self).getPropertyValue(index)
            break

    def makePosSequence(self):
        # FIXME: Private members in OpenDSS
        if self.nPhases != 1:
            Parser.getInstance().setCmdString('Phases=1')
            self.edit()
        super(GICTransformerObjImpl, self).makePosSequence()

    def getG1(self):
        return self.G1

    def setG1(self, g1):
        self.G1 = g1

    def getG2(self):
        return self.G2

    def setG2(self, g2):
        self.G2 = g2

    def getSpecType(self):
        return self.specType

    def setSpecType(self, type):
        self.specType = type
